File: mac/Sources/TokenBar/Services/Log.swift
import Foundation
import os

/// 全项目统一的日志出口。
///
/// 级别按 os_log 的落盘规则挑选，不是随手写的：
/// - `.notice` / `.error` 会持久化到系统日志，事后能用 `log show --last 1h` 回溯；
/// - `.info` / `.debug` 只驻内存缓冲，必须 `log stream` 现场盯着才看得到。
/// 因此"用户过几天来报障、我要能倒查"的事件（定时器 fire、每轮开始结束、
/// 厂商超时、闸门抢占）一律 notice/error；单次请求耗时这类高频细节用 info/debug。
///
/// 排查命令：
///     log stream --predicate 'subsystem == "com.tokenbar.mac"' --level debug --style compact
///     log show --last 1h --predicate 'subsystem == "com.tokenbar.mac"' --info --style compact
///
/// 隐私红线：Logger 对非字面量插值默认按 `<private>` 处理，这是安全的默认值。
/// 只有确定不敏感的值（厂商标识、耗时、HTTP 状态码、触发来源、间隔秒数）才标
/// `privacy: .public`。绝不记录 apiKey / token / accessKeySecret / Cookie /
/// refreshToken 及其任何片段、请求头、请求体；URL 只记 host + path，丢掉 query；
/// error.localizedDescription 保持 private —— 百炼的 aggregateError 会把服务端
/// 回显拼进去，可能带上账号信息。
public enum Log {
    public static let subsystem = "com.tokenbar.mac"

    /// 定时器生命周期：创建、fire、失效自愈
    public static let timer = Logger(subsystem: subsystem, category: "timer")
    /// 刷新轮次：开始、结束、闸门跳过与抢占
    public static let refresh = Logger(subsystem: subsystem, category: "refresh")
    /// 单厂商：耗时、成功、失败原因、超时
    public static let provider = Logger(subsystem: subsystem, category: "provider")
    /// 网络层细节
    public static let net = Logger(subsystem: subsystem, category: "net")
    /// App 生命周期、菜单栏、唤醒
    public static let lifecycle = Logger(subsystem: subsystem, category: "lifecycle")
}

extension URL {
    /// 日志用的安全短名：只保留 host + path，丢掉可能带密钥的 query。
    var logSafeDescription: String {
        "\(host ?? "?")\(path)"
    }
}

File: windows/src/TokenBar/Services/Log.cs
using System;
using System.IO;
using System.Text;

namespace TokenBar.Services
{
    /// <summary>
    /// 轻量文件日志。macOS 端用 os.Logger + `log stream` 排查，Windows 没有等价物，
    /// 因此落到 %APPDATA%\TokenBar\tokenbar.log。
    ///
    /// 存在的理由：定时刷新出问题时，"定时器没触发"和"每轮都失败"这两种成因在界面上
    /// 长得一模一样，没有日志只能靠外部文件时间戳倒推，代价极大。
    ///
    /// 排查用法（PowerShell 实时跟随）：
    ///     Get-Content "$env:APPDATA\TokenBar\tokenbar.log" -Wait -Tail 50
    ///
    /// 隐私红线：绝不记录 apiKey / token / accessKeySecret / Cookie / refreshToken
    /// 及其任何片段、请求头、请求体；URL 只记 host + path，丢掉 query。
    /// </summary>
    public static class Log
    {
        private const long MaxBytes = 2 * 1024 * 1024;

        private static readonly object Gate = new();
        private static readonly string LogPath;
        private static bool _disabled;

        static Log()
        {
            try
            {
                var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
                var dir = Path.Combine(appData, "TokenBar");
                Directory.CreateDirectory(dir);
                LogPath = Path.Combine(dir, "tokenbar.log");
            }
            catch
            {
                LogPath = string.Empty;
                _disabled = true;
            }
        }

        public static void Debug(string category, string message) => Write("DEBUG", category, message);
        public static void Info(string category, string message) => Write("INFO ", category, message);
        public static void Notice(string category, string message) => Write("NOTE ", category, message);
        public static void Warn(string category, string message) => Write("WARN ", category, message);
        public static void Error(string category, string message) => Write("ERROR", category, message);

        private static void Write(string level, string category, string message)
        {
            if (_disabled) return;

            try
            {
                var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff} {level} [{category}] {message}{Environment.NewLine}";
                lock (Gate)
                {
                    // 超过上限就整体轮转一次，保留上一份供对比。日志绝不能把磁盘写满。
                    var info = new FileInfo(LogPath);
                    if (info.Exists && info.Length > MaxBytes)
                    {
                        var backup = LogPath + ".1";
                        if (File.Exists(backup)) File.Delete(backup);
                        File.Move(LogPath, backup);
                    }

                    File.AppendAllText(LogPath, line, Encoding.UTF8);
                }
            }
            catch
            {
                // 日志失败绝不能影响刷新本身：一次失败即永久放弃，避免每轮都抛异常。
                _disabled = true;
            }
        }
    }
}
